vortex-ipc/src/messages/decoder.rs

```diff
@@ -0,0 +1,220 @@
+use std::fmt::{Debug, Formatter};
+use std::sync::Arc;
+
+use bytes::{Buf, BytesMut};
+use flatbuffers::{root, root_unchecked, Follow};
+use itertools::Itertools;
+use vortex_array::{flatbuffers as fba, ArrayData, Context};
+use vortex_buffer::Buffer;
+use vortex_dtype::DType;
+use vortex_error::{vortex_bail, vortex_err, VortexExpect, VortexResult};
+use vortex_flatbuffers::message as fb;
+use vortex_flatbuffers::message::{MessageHeader, MessageVersion};
+
+/// A message decoded from an IPC stream.
+///
+/// Note that the `Array` variant cannot fully decode into an [`ArrayData`] without a [`Context`]
+/// and a [`DType`]. As such, we partially decode into an [`ArrayParts`] and allow the caller to
+/// finish the decoding.
+#[derive(Debug)]
+pub enum DecoderMessage {
+    Array(ArrayParts),
+    Buffer(Buffer),
+    DType(DType),
+}
+
+/// ArrayParts represents a partially decoded Vortex array.
+/// It can be completely decoded calling `into_array_data` with a context and dtype.
+pub struct ArrayParts {
+    row_count: usize,
+    // Typed as fb::Array
+    array_flatbuffer: Buffer,
+    array_flatbuffer_loc: usize,
+    buffers: Vec<Buffer>,
+}
+
+impl Debug for ArrayParts {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ArrayComponents")
+            .field("row_count", &self.row_count)
+            .field("array_flatbuffer", &self.array_flatbuffer.len())
+            .field("buffers", &self.buffers.len())
+            .finish()
+    }
+}
+
+impl ArrayParts {
+    pub fn into_array_data(self, ctx: Arc<Context>, dtype: DType) -> VortexResult<ArrayData> {
+        ArrayData::try_new_viewed(
+            ctx,
+            dtype,
+            self.row_count,
+            self.array_flatbuffer,
+            // SAFETY: ArrayComponents guarantees the buffers are valid.
+            |buf| unsafe { Ok(fba::Array::follow(buf, self.array_flatbuffer_loc)) },
+            self.buffers,
+        )
+    }
+}
+
+#[derive(Default)]
+enum State {
+    #[default]
+    Length,
+    Header(usize),
+    Array(ReadingArray),
+    Buffer(ReadingBuffer),
+}
+
+struct ReadingArray {
+    header: Buffer,
+    buffers_length: usize,
+}
+
+struct ReadingBuffer {
+    length: usize,
+    length_with_padding: usize,
+}
+
+pub enum NextMessage {
```

Comment:
This is effectively the same as `PollRead` from vortex-io, can we do anything to make the names a bit more similar? Or perhaps even do some reuse? It's confusing to see the same concept in multiple places in the codebase under different names

Reply:
Oh weird, it looks like it's defined in `vortex-file`? Happy to use the same name. In Tokio this whole mechanism is known as a "codec": https://docs.rs/tokio-util/latest/tokio_util/codec/index.html#the-decoder-trait

But not sure it should be reusable tbh. LayoutReaders ask for more messages, not more bytes. So it's a similar pattern to lift IO out of a reader, but not much logic to be shared.

Will make the names align though